src/include/ikarus/linearAlgebra/dirichletValues.hh

```diff
@@ -0,0 +1,130 @@
+
+#include <autodiff/forward/real.hpp>
+#include <autodiff/forward/real/eigen.hpp>
+
+#include <ikarus/utils/traits.hh>
+
+namespace Ikarus {
+
+  template <typename Op, typename... Args>
+  concept IsFunctorWithArgs = requires(Op op, Args... args) {
+    op(args...);
+  };
+
+  template <typename Basis_>
+  class DirichletValues {
+  public:
+    using Basis                         = Basis_;
+    using FlagsType                     = std::vector<bool>;
+    static constexpr int worldDimension = Basis::GridView::dimensionworld;
+    using BackendType                   = decltype(Dune::Functions::istlVectorBackend(std::declval<FlagsType&>()));
+    explicit DirichletValues(std::shared_ptr<const Basis>& p_basis)
+        : basis_{p_basis}, dirichletFlagsBackend{dirichletFlags} {
+      dirichletFlagsBackend.resize(*basis_);
+      std::fill(dirichletFlags.begin(), dirichletFlags.end(), false);
+      inhomogeneousBoundaryVectorDummy.setZero(basis_->size());
+    }
+
+    template <typename F>
+    void fixBoundaryDOFs(F&& f) {
+      if constexpr (IsFunctorWithArgs<F, BackendType, typename Basis::MultiIndex>) {
+        auto lambda = [&](auto&& indexGlobal) { f(dirichletFlagsBackend, indexGlobal); };
+        Dune::Functions::forEachBoundaryDOF(*basis_, lambda);
+      } else if constexpr (IsFunctorWithArgs<F, BackendType, int, typename Basis::LocalView>) {
+        auto lambda = [&](auto&& localIndex, auto&& localView) { f(dirichletFlagsBackend, localIndex, localView); };
+        Dune::Functions::forEachBoundaryDOF(*basis_, lambda);
+      } else if constexpr (IsFunctorWithArgs<F, BackendType, int, typename Basis::LocalView,
+                                             typename Basis::GridView::Intersection>) {
+        auto lambda = [&](auto&& localIndex, auto&& localView, auto&& intersection) {
+          f(dirichletFlagsBackend, localIndex, localView, intersection);
+        };
+        Dune::Functions::forEachBoundaryDOF(*basis_, lambda);
+      }
+    }
+
+    template <typename F>
+    void fixDOFs(F&& f) {
+      f(*basis_, dirichletFlags);
+    }
+
+    const auto& basis() const { return *basis_; }
+
+    [[nodiscard]] bool isConstrained(std::size_t i) const { return dirichletFlags[i]; }
+
+    auto fixedSize() const { return std::ranges::count(dirichletFlags, true); }
+    auto size() const { return dirichletFlags.size(); }
+    auto& vector() { return dirichletFlags; }
+
+    template <typename F>
+    void storeInhomogeneousBoundaryCondition(F&& f) {
+      auto derivativeLambda = [&](const auto& globalCoord, const double& lambda) {
+        autodiff::real lambdadual = lambda;
+        lambdadual[1]             = 1;  // Setting the derivative in lambda direction to 1
+        return derivative(f(globalCoord, lambdadual));
+      };
+      dirichletFunctions.push_back({f, derivativeLambda});
+    }
+
+    void evaluateInhomogeneousBoundaryCondition(Eigen::VectorXd& xIh, const double& lambda) {
```

Comment:
Can be renamed to evaluateInhomogeneousBoundaryConditionValue to have the notion of value and derivative and/or BoundaryCondition can be abbreviated as BC?